File: frontend/src/types/journal.ts
import type { TradeDirection, TradeOutcome } from "./common";

export interface JournalEntry {
  id: number;
  created_at: string;
  updated_at: string;
  setup_id: number | null;
  account_id: number | null;
  mt5_ticket: number | null;
  symbol: string;
  direction: TradeDirection;
  entry_price: number;
  exit_price: number | null;
  stop_loss: number;
  take_profit: number;
  lot_size: number;
  risk_percent: number | null;
  r_multiple: number | null;
  profit_loss: number | null;
  profit_loss_percent: number | null;
  outcome: TradeOutcome;
  opened_at: string;
  closed_at: string | null;
  ai_summary: string | null;
  trader_notes: string | null;
}

export interface JournalEntryCreate {
  setup_id?: number | null;
  account_id?: number | null;
  mt5_ticket?: number | null;
  symbol: string;
  direction: TradeDirection;
  entry_price: number;
  stop_loss: number;
  take_profit: number;
  lot_size: number;
  risk_percent?: number | null;
  opened_at: string;
  trader_notes?: string | null;
}

export interface JournalEntryUpdate {
  exit_price?: number | null;
  closed_at?: string | null;
  outcome?: TradeOutcome | null;
  profit_loss?: number | null;
  profit_loss_percent?: number | null;
  r_multiple?: number | null;
  ai_summary?: string | null;
  trader_notes?: string | null;
}

export interface ScreenshotAnalysis {
  id: number;
  created_at: string;
  updated_at: string;
  journal_entry_id: number | null;
  file_path: string;
  symbol_hint: string | null;
  detected_trend: string | null;
  detected_support: number[];
  detected_resistance: number[];
  suggested_entry: number | null;
  suggested_stop_loss: number | null;
  suggested_take_profit: number | null;
  mistakes: string[];
  risk_notes: string | null;
  full_report: string;
}

export interface WinRateReport {
  total_trades: number;
  wins: number;
  losses: number;
  breakevens: number;
  win_rate_percent: number;
  average_r_multiple: number | null;
  total_profit_loss: number;
}

export interface PeriodReport {
  period_start: string;
  period_end: string;
  trades: JournalEntry[];
  win_rate: WinRateReport;
  net_profit_loss: number;
  best_trade: JournalEntry | null;
  worst_trade: JournalEntry | null;
}

File: backend/app/schemas/journal.py
"""Pydantic schemas for the trading journal and screenshot analysis."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from app.core.constants import TradeDirection, TradeOutcome
from app.schemas.common import TimestampedSchema


class JournalEntryCreate(BaseModel):
    setup_id: int | None = None
    account_id: int | None = None
    mt5_ticket: int | None = None
    symbol: str
    direction: TradeDirection
    entry_price: float
    stop_loss: float
    take_profit: float
    lot_size: float = Field(..., gt=0)
    risk_percent: float | None = None
    opened_at: datetime
    trader_notes: str | None = None


class JournalEntryUpdate(BaseModel):
    exit_price: float | None = None
    closed_at: datetime | None = None
    outcome: TradeOutcome | None = None
    profit_loss: float | None = None
    profit_loss_percent: float | None = None
    r_multiple: float | None = None
    ai_summary: str | None = None
    trader_notes: str | None = None


class JournalEntryRead(TimestampedSchema):
    setup_id: int | None = None
    account_id: int | None = None
    mt5_ticket: int | None = None
    symbol: str
    direction: TradeDirection
    entry_price: float
    exit_price: float | None = None
    stop_loss: float
    take_profit: float
    lot_size: float
    risk_percent: float | None = None
    r_multiple: float | None = None
    profit_loss: float | None = None
    profit_loss_percent: float | None = None
    outcome: TradeOutcome
    opened_at: datetime
    closed_at: datetime | None = None
    ai_summary: str | None = None
    trader_notes: str | None = None


class ScreenshotAnalysisRead(TimestampedSchema):
    journal_entry_id: int | None = None
    file_path: str
    symbol_hint: str | None = None
    detected_trend: str | None = None
    detected_support: list[float]
    detected_resistance: list[float]
    suggested_entry: float | None = None
    suggested_stop_loss: float | None = None
    suggested_take_profit: float | None = None
    mistakes: list[str]
    risk_notes: str | None = None
    full_report: str


class WinRateReport(BaseModel):
    total_trades: int
    wins: int
    losses: int
    breakevens: int
    win_rate_percent: float
    average_r_multiple: float | None = None
    total_profit_loss: float


class PeriodReport(BaseModel):
    period_start: datetime
    period_end: datetime
    trades: list[JournalEntryRead]
    win_rate: WinRateReport
    net_profit_loss: float
    best_trade: JournalEntryRead | None = None
    worst_trade: JournalEntryRead | None = None
